clash-prelude/src/Clash/Sized/Vector/ToTuple.hs

```diff
@@ -0,0 +1,95 @@
+{-|
+Copyright  :  (C) 2024, Google LLC
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Tooling to work around incomplete pattern match warnings produced by newer
+GHC versions. See 'vecToTuple' for more information and examples.
+-}
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+
+-- Purpose of this module
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+
+-- For debugging:
+-- {-# OPTIONS_GHC -ddump-splices #-}
+
+module Clash.Sized.Vector.ToTuple (VecToTuple(..)) where
+
+import Clash.CPP
+import Clash.Sized.Vector
+import Clash.Sized.Vector.ToTuple.TH (vecToTupleInstances)
+
+import Data.Tagged (Tagged(..))
+
+#if MIN_VERSION_base(4,18,0)
+import Data.Tuple (Solo(MkSolo))
+#elif MIN_VERSION_base(4,16,0)
+import Data.Tuple (Solo(Solo))
+#endif
+
+{- $setup
+>>> :set -XMonoLocalBinds -XGADTs
+>>> import Clash.Sized.Vector
+-}
+
+class VecToTuple a where
+  type TupType a = r | r -> a
+
+  -- | Given a vector with three elements:
+  --
+  -- >>> myVec = (1 :> 2 :> 3 :> Nil) :: Vec 3 Int
+  --
+  -- The following would produce a warning even though we can be sure
+  -- no other pattern can ever apply:
+  --
+  -- >>> (a :> b :> c :> Nil) = myVec
+  --
+  -- 'vecToTuple' can be used to work around the warning:
+  --
+  -- >>> (a, b, c) = vecToTuple myVec
+  --
+  -- Of course, you will still get an error if you try to match a vector of the
+  -- wrong length:
+  --
+  -- >>> (a, b, c, d) = vecToTuple myVec
+  -- ...
+#if __GLASGOW_HASKELL__ > 900
+  --     • Couldn't match type: (a, b, c, d)
+  --                      with: (Int, Int, Int)
+#else
+  --     • Couldn't match type: (Int, Int, Int)
+  --                             ^
+  --                      with: (a, b, c, d)
+#endif
+  -- ...
+  --
+  vecToTuple ::  a -> TupType a
+
+instance VecToTuple (Vec 0 a) where
+  type TupType (Vec 0 a) = Tagged a ()
+  vecToTuple Nil = Tagged ()
+
+#if MIN_VERSION_base(4,18,0)
+instance VecToTuple (Vec 1 a) where
+  type TupType (Vec 1 a) = Solo a
```

Comment:
Would it make sense to put haddock comment on the `Vec 1` instance that you could also use `head` instead?